src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests.cs

```diff
@@ -5043,6 +5043,228 @@ class C<T>
+            attribute.VerifyValue(0, TypedConstantKind.Primitive, "one");
+            Assert.Equal(@"MyAttribute(""one"")", attribute.ToString());
+
+            CompileAndVerify(source, expectedOutput: "Hello World");
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void TestMainAttributes_IL()
+        {
+            var source = CreateCompilation(@"
+using System;
+
+[main: My(""one"")]
+Console.WriteLine(""Hello World"");
+
+public class MyAttribute : Attribute { public MyAttribute(string name) {} }
+");
+
+            CompileAndVerify(source, expectedOutput: "Hello World")
+                .VerifyTypeIL("Program", @"
+.class private auto ansi beforefieldinit Program
+	extends [netstandard]System.Object
+{
+	.custom instance void [netstandard]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = (
+		01 00 00 00
+	)
+	// Methods
+	.method private hidebysig static 
+		void '<Main>$' (
+			string[] args
+		) cil managed 
+	{
+		.custom instance void MyAttribute::.ctor(string) = (
+			01 00 03 6f 6e 65 00 00
+		)
+		// Method begins at RVA 0x2050
+		// Code size 11 (0xb)
+		.maxstack 8
+		.entrypoint
+		IL_0000: ldstr ""Hello World""
+
+        IL_0005: call void[netstandard]System.Console::WriteLine(string)
+
+        IL_000a: ret
+
+    } // end of method Program::'<Main>$'
+	.method public hidebysig specialname rtspecialname
+        instance void .ctor() cil managed
+        {
+		// Method begins at RVA 0x205c
+		// Code size 7 (0x7)
+		.maxstack 8
+		IL_0000: ldarg.0
+		IL_0001: call instance void [netstandard] System.Object::.ctor()
+         IL_0006: ret
+    } // end of method Program::.ctor
+} // end of class Program
+");
+        }
+
+        [Fact]
+        public void TestMultipleMainAttributes()
+        {
+            var source = CreateCompilation(@"
+using System;
+
+[main: My(""one"")]
+[main: My(""two"")]
+[main: My(""three"")]
+Console.WriteLine(""Hello World"");
+
+[AttributeUsage(AttributeTargets.All, AllowMultiple = true)]
+public class MyAttribute : Attribute { public MyAttribute(string name) {} }
+");
+
+            var mainAttrs = source.GetEntryPoint(default).GetAttributes();
+
+            Assert.Collection(mainAttrs,
+                verifyAttribute("one"),
+                verifyAttribute("two"),
+                verifyAttribute("three"));
+
+            CompileAndVerify(source, expectedOutput: "Hello World");
+
+            static Action<CSharpAttributeData> verifyAttribute(string expectedValue) => (attribute) =>
+            {
+                attribute.VerifyValue(0, TypedConstantKind.Primitive, expectedValue);
+                Assert.Equal($@"MyAttribute(""{expectedValue}"")", attribute.ToString());
+            };
+        }
+
+        [Fact]
+        public void TestMainAttributesWithAssemblyAndModuleAttributes()
+        {
+            var source = CreateCompilation(@"
+using System;
+
+[main: My(""one"")]
+[assembly: My(""two"")]
```

Comment:
I think it will be good to test where an attribute without an explicit target is going to end up.